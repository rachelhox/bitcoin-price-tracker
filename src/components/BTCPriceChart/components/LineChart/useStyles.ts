import { css, useTheme } from '@mui/material';

const useStyles = () => {
  const theme = useTheme();
  return {
    root: css({
      margin: 'auto',
      'svg:not(:root)': {
        overflow: 'visible',
      },
      [theme.breakpoints.up('laptop')]: {
        width: '70%',
      },
    }),
  };
};

export default useStyles;
