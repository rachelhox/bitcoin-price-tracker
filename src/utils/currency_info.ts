interface IStringIndex extends Record<string, any> {}

/**
 * Takes the key name and returns the location. Will return null if undefined
 */

const currencies: IStringIndex = {
  usd: {
    name: 'US Dollar',
    key: 'usd',
    unit: '$',
    type: 'fiat',
  },
  aed: {
    name: 'United Arab Emirates Dirham',
    key: 'aed',
    unit: 'DH',
    type: 'fiat',
  },
  eur: {
    name: 'Euro',
    key: 'eur',
    unit: '€',
    type: 'fiat',
  },
  aud: {
    name: 'Australian Dollar',
    key: 'aud',
    unit: 'A$',
    type: 'fiat',
  },
  cad: {
    name: 'Canadian Dollar',
    key: 'cad',
    unit: 'CA$',
    type: 'fiat',
  },
  chf: {
    name: 'Swiss Franc',
    key: 'chf',
    unit: 'Fr.',
    type: 'fiat',
  },
  gbp: {
    name: 'British Pound Sterling',
    key: 'gbp',
    unit: '£',
    type: 'fiat',
  },
  hkd: {
    name: 'Hong Kong Dollar',
    key: 'hkd',
    unit: 'HK$',
    type: 'fiat',
  },
  sgd: {
    name: 'Singapore Dollar',
    key: 'sgd',
    unit: 'S$',
    type: 'fiat',
  },
  cny: {
    name: 'Chinese Yuan',
    key: 'cny',
    unit: '¥',
    type: 'fiat',
  },
  jpy: {
    name: 'Japanese Yen',
    key: 'jpy',
    unit: '¥',
    type: 'fiat',
  },
  inr: {
    name: 'Indian Rupee',
    key: 'inr',
    unit: '₹',
    type: 'fiat',
  },
  krw: {
    name: 'South Korean Won',
    key: 'krw',
    unit: '₩',
    type: 'fiat',
  },
};

export const getCurrencyInfo = (key: string | number) => {
  return currencies[key] ?? {};
};
