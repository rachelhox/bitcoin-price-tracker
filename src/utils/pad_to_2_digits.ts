/* eslint-disable */

export const padTo2Digits = (num: any) => {
  return String(num).padStart(2, '0');
};
