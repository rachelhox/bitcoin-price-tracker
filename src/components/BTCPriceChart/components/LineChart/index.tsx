/* eslint-disable no-unused-vars */
import { Box, useTheme, useMediaQuery } from '@mui/material';
import {
  VictoryChart,
  VictoryAxis,
  VictoryLine,
  VictoryTheme,
  createContainer,
  VictoryTooltip,
} from 'victory';
import { useGeneralContext } from '@contexts';
import { convertToMoney, padTo2Digits } from '@utils';
import { NoSSR } from '@components';
import useStyles from './useStyles';

interface LineChartProp {
  // eslint-disable-next-line no-undef
  lineChartData: PriceProps[];
}

const VictoryZoomVoronoiContainer: any = createContainer('zoom', 'voronoi');

const LineChart = ({ lineChartData }: LineChartProp) => {
  const styles = useStyles();
  const theme = useTheme();
  const { currency } = useGeneralContext();
  const onlyLargeScreen = useMediaQuery(theme.breakpoints.up('laptop'));
  return (
    <NoSSR>
      <Box css={styles.root}>
        <VictoryChart
          theme={VictoryTheme.material}
          domainPadding={{ x: 10 }}
          padding={{ left: 90, top: 10, right: 0, bottom: 30 }}
          scale={{ x: 'time' }}
          containerComponent={
            <VictoryZoomVoronoiContainer
              labels={({ datum }: any) =>
                `time: ${padTo2Digits(datum.x.getHours())}:${padTo2Digits(
                  datum.x.getMinutes()
                )}\
                    price: ${currency.unit}${convertToMoney(datum.y)}`
              }
              labelComponent={
                <VictoryTooltip
                  cornerRadius={8}
                  flyoutStyle={{
                    fill: theme.palette.common.black,
                    color: theme.palette.common.white,
                    stroke: theme.palette.common.white,
                  }}
                  style={{
                    fill: theme.palette.common.white,
                    fontSize: 8,
                  }}
                />
              }
              disable={!onlyLargeScreen}
            />
          }
        >
          <VictoryAxis tickFormat={(t) => `${t.getDate()}/${t.getMonth()}`} />
          <VictoryAxis
            dependentAxis
            tickFormat={(t) => [`${currency.unit}${convertToMoney(t)}`]}
          />
          <VictoryLine
            interpolation="natural"
            data={lineChartData}
            style={{
              data: { stroke: theme.palette.custom.yellow, strokeWidth: 1 },
            }}
          />
        </VictoryChart>
      </Box>
    </NoSSR>
  );
};

export default LineChart;
