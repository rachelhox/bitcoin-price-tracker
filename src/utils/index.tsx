import { getCurrencyInfo } from './currency_info';
import { convertToMoney } from './convert_to_money';
import { padTo2Digits } from './pad_to_2_digits';

export { getCurrencyInfo, convertToMoney, padTo2Digits };
