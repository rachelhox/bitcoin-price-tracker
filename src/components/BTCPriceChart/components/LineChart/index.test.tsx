import renderer from 'react-test-renderer';
import { MockTheme } from '@tests/utils';
import LineChart from '.';

// ==================================
// mocks
// ==================================
const mockLineChartData = [
  { x: new Date(1660291413194), y: 137359.8917588185 },
];

jest.mock('@components', () => ({
  NoSSR: (props: any) => <div id="NoSSR" {...props} />,
}));

// ==================================
// unit tests
// ==================================
describe('component: LineChart', () => {
  it('matches snapshot', () => {
    const component = renderer.create(
      <MockTheme>
        <LineChart lineChartData={mockLineChartData} />
      </MockTheme>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
