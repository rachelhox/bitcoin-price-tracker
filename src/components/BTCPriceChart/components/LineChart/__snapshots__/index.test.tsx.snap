// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: LineChart matches snapshot 1`] = `
<div
  id="NoSSR"
>
  <div
    className="MuiBox-root css-p1qljd-LineChart"
  >
    <div
      className="VictoryContainer"
      style={
        {
          "height": "100%",
          "pointerEvents": "none",
          "position": "relative",
          "touchAction": "none",
          "userSelect": "none",
          "width": "100%",
        }
      }
    >
      <svg
        height={350}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        onWheel={[Function]}
        role="img"
        style={
          {
            "height": "100%",
            "pointerEvents": "all",
            "width": "100%",
          }
        }
        viewBox="0 0 350 350"
        width={350}
      >
        <g
          role="presentation"
        >
          <line
            role="presentation"
            shapeRendering="auto"
            style={
              {
                "fill": "transparent",
                "stroke": "#90A4AE",
                "strokeLinecap": "round",
                "strokeLinejoin": "round",
                "strokeWidth": 2,
              }
            }
            vectorEffect="non-scaling-stroke"
            x1={90}
            x2={350}
            y1={320}
            y2={320}
          />
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={90}
              y1={320}
              y2={10}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={90}
              y1={320}
              y2={325}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-0-tickLabels-0"
              x={90}
              y={343.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="middle"
                x={90}
              >
                12/7
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={176.66666666666666}
              x2={176.66666666666666}
              y1={320}
              y2={10}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={176.66666666666666}
              x2={176.66666666666666}
              y1={320}
              y2={325}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-0-tickLabels-1"
              x={176.66666666666666}
              y={343.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="middle"
                x={176.66666666666666}
              >
                12/7
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={263.3333333333333}
              x2={263.3333333333333}
              y1={320}
              y2={10}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={263.3333333333333}
              x2={263.3333333333333}
              y1={320}
              y2={325}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-0-tickLabels-2"
              x={263.3333333333333}
              y={343.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="middle"
                x={263.3333333333333}
              >
                12/7
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={350}
              x2={350}
              y1={320}
              y2={10}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={350}
              x2={350}
              y1={320}
              y2={325}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-0-tickLabels-3"
              x={350}
              y={343.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="middle"
                x={350}
              >
                12/7
              </tspan>
            </text>
          </g>
        </g>
        <g
          role="presentation"
        >
          <line
            role="presentation"
            shapeRendering="auto"
            style={
              {
                "fill": "transparent",
                "stroke": "#90A4AE",
                "strokeLinecap": "round",
                "strokeLinejoin": "round",
                "strokeWidth": 2,
              }
            }
            vectorEffect="non-scaling-stroke"
            x1={90}
            x2={90}
            y1={10}
            y2={320}
          />
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={350}
              y1={320}
              y2={320}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={85}
              y1={320}
              y2={320}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-1-tickLabels-0"
              x={77}
              y={324.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="end"
                x={77}
              >
                $137,360
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={350}
              y1={268.33333333333337}
              y2={268.33333333333337}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={85}
              y1={268.33333333333337}
              y2={268.33333333333337}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-1-tickLabels-1"
              x={77}
              y={272.59333333333336}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="end"
                x={77}
              >
                $137,360
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={350}
              y1={165}
              y2={165}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={85}
              y1={165}
              y2={165}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-1-tickLabels-2"
              x={77}
              y={169.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="end"
                x={77}
              >
                $137,360
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={350}
              y1={61.66666666666666}
              y2={61.66666666666666}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={85}
              y1={61.66666666666666}
              y2={61.66666666666666}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-1-tickLabels-3"
              x={77}
              y={65.92666666666666}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="end"
                x={77}
              >
                $137,360
              </tspan>
            </text>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "none",
                  "pointerEvents": "painted",
                  "stroke": "#ECEFF1",
                  "strokeDasharray": "10, 5",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={350}
              y1={10}
              y2={10}
            />
            <line
              role="presentation"
              shapeRendering="auto"
              style={
                {
                  "fill": "transparent",
                  "size": 5,
                  "stroke": "#90A4AE",
                  "strokeLinecap": "round",
                  "strokeLinejoin": "round",
                  "strokeWidth": 1,
                }
              }
              vectorEffect="non-scaling-stroke"
              x1={90}
              x2={85}
              y1={10}
              y2={10}
            />
            <text
              direction="inherit"
              dx={0}
              id="chart-axis-1-tickLabels-4"
              x={77}
              y={14.26}
            >
              <tspan
                dx={0}
                dy={0}
                style={
                  {
                    "fill": "#455A64",
                    "fontFamily": "'Helvetica Neue', 'Helvetica', sans-serif",
                    "fontSize": 12,
                    "letterSpacing": "normal",
                    "padding": 8,
                    "stroke": "transparent",
                    "strokeWidth": 0,
                  }
                }
                textAnchor="end"
                x={77}
              >
                $137,360
              </tspan>
            </text>
          </g>
        </g>
        <g
          clipPath="url(#victory-clip-4)"
          style={
            {
              "height": "100%",
              "userSelect": "none",
              "width": "100%",
            }
          }
        >
          <defs>
            <clipPath
              id="victory-clip-4"
            >
              <rect
                height={310}
                vectorEffect="non-scaling-stroke"
                width={260}
                x={90}
                y={10}
              />
            </clipPath>
          </defs>
          <path
            d={null}
            role="presentation"
            shapeRendering="auto"
            style={
              {
                "fill": "transparent",
                "opacity": 1,
                "stroke": "#F4D933",
                "strokeWidth": 1,
              }
            }
          />
        </g>
      </svg>
      <div
        style={
          {
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
            "zIndex": 99,
          }
        }
      >
        <svg
          height={350}
          style={
            {
              "height": "100%",
              "overflow": "visible",
              "width": "100%",
            }
          }
          viewBox="0 0 350 350"
          width={350}
        />
      </div>
    </div>
  </div>
</div>
`;
